Special characters in table name cause exception

```
2023-01-14 00:17:35,842 ERROR [clickhouse-sink.cdc.structured-trait-updates.v0|task-0] Got exception when running DESCRIBE TABLE default.pocus.cdc.structured-trait-updates.v0 (com.clickhouse.kafka.connect.sink.db.helper.ClickHouseHelperClient) [task-thread-clickhouse-sink.cdc.structured-trait-updates.v0-0]
com.clickhouse.client.ClickHouseException: Code: 62. DB::Exception: Syntax error: failed at position 44 ('-'): -trait-updates.v0. Expected one of: token, Dot. (SYNTAX_ERROR) (version 22.9.3.18 (official build))
, server ClickHouseNode [uri=http://clickhouse:8123/default]
	at com.clickhouse.client.ClickHouseException.of(ClickHouseException.java:148)
	at com.clickhouse.client.AbstractClient.lambda$execute$0(AbstractClient.java:291)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.IOException: Code: 62. DB::Exception: Syntax error: failed at position 44 ('-'): -trait-updates.v0. Expected one of: token, Dot. (SYNTAX_ERROR) (version 22.9.3.18 (official build))

	at com.clickhouse.client.http.HttpUrlConnectionImpl.checkResponse(HttpUrlConnectionImpl.java:187)
	at com.clickhouse.client.http.HttpUrlConnectionImpl.post(HttpUrlConnectionImpl.java:278)
	at com.clickhouse.client.http.ClickHouseHttpClient.send(ClickHouseHttpClient.java:101)
	at com.clickhouse.client.AbstractClient.sendAsync(AbstractClient.java:156)
	at com.clickhouse.client.AbstractClient.lambda$execute$0(AbstractClient.java:289)
	... 4 more
```

It looks like the `DESCRIBE TABLE` call is not back-quoting the table name, which is causing a SQL syntax error when the table name is not strictly alphanumeric.

--- a/src/main/java/com/clickhouse/kafka/connect/sink/db/ClickHouseWriter.java
+++ b/src/main/java/com/clickhouse/kafka/connect/sink/db/ClickHouseWriter.java
@@ -95,6 +95,9 @@ public class ClickHouseWriter implements DBWriter{
         isBinary = binary;
     }
 
+    private String escapeTopicName(String topic) {
+        return String.format("`%s`", topic);
+    }
     // TODO: we need to refactor that
     private String convertHelper(Object v) {
         if (v instanceof List) {
@@ -263,7 +266,7 @@ public class ClickHouseWriter implements DBWriter{
         Record first = records.get(0);
         String topic = first.getTopic();
         LOGGER.info(String.format("Number of records to insert %d to table name %s", batchSize, topic));
-        Table table = this.mapping.get(topic);
+        Table table = this.mapping.get(escapeTopicName(topic));
         if (table == null) {
             //TODO to pick the correct exception here
             throw new RuntimeException(String.format("Table %s does not exists", topic));
@@ -337,7 +340,7 @@ public class ClickHouseWriter implements DBWriter{
         Record first = records.get(0);
         String topic = first.getTopic();
         LOGGER.info(String.format("Number of records to insert %d to table name %s", batchSize, topic));
-        Table table = this.mapping.get(topic);
+        Table table = this.mapping.get(escapeTopicName(topic));
         if (table == null) {
             //TODO to pick the correct exception here
             throw new RuntimeException(String.format("Table %s does not exists", topic));
@@ -354,7 +357,7 @@ public class ClickHouseWriter implements DBWriter{
         try (ClickHouseClient client = ClickHouseClient.newInstance(ClickHouseProtocol.HTTP)) {
             ClickHouseRequest.Mutation request = client.connect(chc.getServer())
                     .write()
-                    .table(topic)
+                    .table(table.getName())
                     .format(ClickHouseFormat.JSONEachRow)
                     // this is needed to get meaningful response summary
                     .set("insert_quorum", 2)
@@ -423,7 +426,7 @@ public class ClickHouseWriter implements DBWriter{
         LOGGER.info(String.format("Number of records to insert %d to table name %s", batchSize, topic));
         // Build the insert SQL
         StringBuffer sb = new StringBuffer();
-        sb.append(String.format("INSERT INTO %s ", topic));
+        sb.append(String.format("INSERT INTO %s ", escapeTopicName(topic)));
         sb.append(extractFields(first.getFields(), "(", ")", ",", ""));
         sb.append(" VALUES ");
         LOGGER.debug("sb {}", sb);

--- a/src/main/java/com/clickhouse/kafka/connect/sink/db/helper/ClickHouseHelperClient.java
+++ b/src/main/java/com/clickhouse/kafka/connect/sink/db/helper/ClickHouseHelperClient.java
@@ -129,7 +129,7 @@ public class ClickHouseHelperClient {
     public Table describeTable(String tableName) {
         if (tableName.startsWith(".inner"))
             return null;
-        String describeQuery = String.format("DESCRIBE TABLE %s.%s", this.database, tableName);
+        String describeQuery = String.format("DESCRIBE TABLE `%s`.`%s`", this.database, tableName);
 
         try (ClickHouseClient client = ClickHouseClient.newInstance(ClickHouseProtocol.HTTP);
              ClickHouseResponse response = client.connect(server) // or client.connect(endpoints)

--- a/src/main/java/com/clickhouse/kafka/connect/sink/db/mapping/Table.java
+++ b/src/main/java/com/clickhouse/kafka/connect/sink/db/mapping/Table.java
@@ -17,7 +17,7 @@ public class Table {
     }
 
     public String getName() {
-        return name;
+        return String.format("`%s`", name);
     }
 
     public void addColumn(Column column) {

--- a/src/test/java/com/clickhouse/kafka/connect/sink/ClickHouseSinkTaskSchemalessTest.java
+++ b/src/test/java/com/clickhouse/kafka/connect/sink/ClickHouseSinkTaskSchemalessTest.java
@@ -53,7 +53,7 @@ public class ClickHouseSinkTaskSchemalessTest {
 
 
     private void dropTable(ClickHouseHelperClient chc, String tableName) {
-        String dropTable = String.format("DROP TABLE IF EXISTS %s", tableName);
+        String dropTable = String.format("DROP TABLE IF EXISTS `%s`", tableName);
         try (ClickHouseClient client = ClickHouseClient.newInstance(ClickHouseProtocol.HTTP);
              ClickHouseResponse response = client.connect(chc.getServer()) // or client.connect(endpoints)
                      // you'll have to parse response manually if using a different format
@@ -88,7 +88,7 @@ public class ClickHouseSinkTaskSchemalessTest {
     }
 
     private int countRows(ClickHouseHelperClient chc, String topic) {
-        String queryCount = String.format("select count(*) from %s", topic);
+        String queryCount = String.format("select count(*) from `%s`", topic);
         try (ClickHouseClient client = ClickHouseClient.newInstance(ClickHouseProtocol.HTTP);
              ClickHouseResponse response = client.connect(chc.getServer()) // or client.connect(endpoints)
                      // you'll have to parse response manually if using a different format
@@ -301,6 +301,32 @@ public class ClickHouseSinkTaskSchemalessTest {
         String topic = "schemaless_map_table_test";
         dropTable(chc, topic);
         createTable(chc, topic, "CREATE TABLE %s ( `off16` Int16, map_string_string Map(String, String), map_string_int64 Map(String, Int64), map_int64_string Map(Int64, String)  ) Engine = MergeTree ORDER BY off16");
+        // https://github.com/apache/kafka/blob/trunk/connect/api/src/test/java/org/apache/kafka/connect/data/StructTest.java#L95-L98
+        Collection<SinkRecord> sr = createMapType(topic, 1);
+
+        ClickHouseSinkTask chst = new ClickHouseSinkTask();
+        chst.start(props);
+        chst.put(sr);
+        chst.stop();
+        assertEquals(sr.size(), countRows(chc, topic));
+    }
+
+    @Test
+    // https://github.com/ClickHouse/clickhouse-kafka-connect/issues/38
+    public void specialCharTest() {
+        Map<String, String> props = new HashMap<>();
+        props.put(ClickHouseSinkConnector.HOSTNAME, db.getHost());
+        props.put(ClickHouseSinkConnector.PORT, db.getFirstMappedPort().toString());
+        props.put(ClickHouseSinkConnector.DATABASE, "default");
+        props.put(ClickHouseSinkConnector.USERNAME, db.getUsername());
+        props.put(ClickHouseSinkConnector.PASSWORD, db.getPassword());
+        props.put(ClickHouseSinkConnector.SSL_ENABLED, "false");
+
+        ClickHouseHelperClient chc = createClient(props);
+
+        String topic = "special-char-table-test";
+        dropTable(chc, topic);
+        createTable(chc, topic, "CREATE TABLE `%s` ( `off16` Int16, map_string_string Map(String, String), map_string_int64 Map(String, Int64), map_int64_string Map(Int64, String)  ) Engine = MergeTree ORDER BY off16");
         // https://github.com/apache/kafka/blob/trunk/connect/api/src/test/java/org/apache/kafka/connect/data/StructTest.java#L95-L98
         Collection<SinkRecord> sr = createMapType(topic, 1);
 

--- a/src/test/java/com/clickhouse/kafka/connect/sink/ClickHouseSinkTaskSchemalessTest.java
+++ b/src/test/java/com/clickhouse/kafka/connect/sink/ClickHouseSinkTaskSchemalessTest.java
@@ -313,7 +313,7 @@ public class ClickHouseSinkTaskSchemalessTest {
 
     @Test
     // https://github.com/ClickHouse/clickhouse-kafka-connect/issues/38
-    public void specialCharTest() {
+    public void specialCharTableNameTest() {
         Map<String, String> props = new HashMap<>();
         props.put(ClickHouseSinkConnector.HOSTNAME, db.getHost());
         props.put(ClickHouseSinkConnector.PORT, db.getFirstMappedPort().toString());

--- a/src/test/java/com/clickhouse/kafka/connect/sink/ClickHouseSinkTaskWithSchemaTest.java
+++ b/src/test/java/com/clickhouse/kafka/connect/sink/ClickHouseSinkTaskWithSchemaTest.java
@@ -56,7 +56,7 @@ public class ClickHouseSinkTaskWithSchemaTest {
 
 
     private void dropTable(ClickHouseHelperClient chc, String tableName) {
-        String dropTable = String.format("DROP TABLE IF EXISTS %s", tableName);
+        String dropTable = String.format("DROP TABLE IF EXISTS `%s`", tableName);
         try (ClickHouseClient client = ClickHouseClient.newInstance(ClickHouseProtocol.HTTP);
              ClickHouseResponse response = client.connect(chc.getServer()) // or client.connect(endpoints)
                      // you'll have to parse response manually if using a different format
@@ -91,7 +91,7 @@ public class ClickHouseSinkTaskWithSchemaTest {
     }
 
     private int countRows(ClickHouseHelperClient chc, String topic) {
-        String queryCount = String.format("select count(*) from %s", topic);
+        String queryCount = String.format("select count(*) from `%s`", topic);
         try (ClickHouseClient client = ClickHouseClient.newInstance(ClickHouseProtocol.HTTP);
              ClickHouseResponse response = client.connect(chc.getServer()) // or client.connect(endpoints)
                      // you'll have to parse response manually if using a different format
@@ -302,6 +302,33 @@ public class ClickHouseSinkTaskWithSchemaTest {
         createTable(chc, topic, "CREATE TABLE %s ( `off16` Int16, `arr` Array(String), `arr_int8` Array(Int8), `arr_int16` Array(Int16), `arr_int32` Array(Int32), `arr_int64` Array(Int64), `arr_float32` Array(Float32), `arr_float64` Array(Float64), `arr_bool` Array(Bool)  ) Engine = MergeTree ORDER BY off16");
         createTable(chc, topic + "mate", "CREATE MATERIALIZED VIEW %s ( `off16` Int16 ) Engine = MergeTree ORDER BY `off16` POPULATE AS SELECT off16 FROM m_array_string_table_test ");
         Collection<SinkRecord> sr = createArrayType(topic, 1);
+
+        ClickHouseSinkTask chst = new ClickHouseSinkTask();
+        chst.start(props);
+        chst.put(sr);
+        chst.stop();
+
+        assertEquals(sr.size(), countRows(chc, topic));
+    }
+
+    @Test
+    // https://github.com/ClickHouse/clickhouse-kafka-connect/issues/38
+    public void specialCharTest() {
+        Map<String, String> props = new HashMap<>();
+        props.put(ClickHouseSinkConnector.HOSTNAME, db.getHost());
+        props.put(ClickHouseSinkConnector.PORT, db.getFirstMappedPort().toString());
+        props.put(ClickHouseSinkConnector.DATABASE, "default");
+        props.put(ClickHouseSinkConnector.USERNAME, db.getUsername());
+        props.put(ClickHouseSinkConnector.PASSWORD, db.getPassword());
+        props.put(ClickHouseSinkConnector.SSL_ENABLED, "false");
+
+        ClickHouseHelperClient chc = createClient(props);
+
+        String topic = "special-char-table-test";
+        dropTable(chc, topic);
+        createTable(chc, topic, "CREATE TABLE `%s` ( `off16` Int16, map_string_string Map(String, String), map_string_int64 Map(String, Int64), map_int64_string Map(Int64, String)  ) Engine = MergeTree ORDER BY off16");
+        // https://github.com/apache/kafka/blob/trunk/connect/api/src/test/java/org/apache/kafka/connect/data/StructTest.java#L95-L98
+        Collection<SinkRecord> sr = createMapType(topic, 1);
 
         ClickHouseSinkTask chst = new ClickHouseSinkTask();
         chst.start(props);

--- a/src/test/java/com/clickhouse/kafka/connect/sink/ClickHouseSinkTaskWithSchemaTest.java
+++ b/src/test/java/com/clickhouse/kafka/connect/sink/ClickHouseSinkTaskWithSchemaTest.java
@@ -313,7 +313,7 @@ public class ClickHouseSinkTaskWithSchemaTest {
 
     @Test
     // https://github.com/ClickHouse/clickhouse-kafka-connect/issues/38
-    public void specialCharTest() {
+    public void specialCharTableNameTest() {
         Map<String, String> props = new HashMap<>();
         props.put(ClickHouseSinkConnector.HOSTNAME, db.getHost());
         props.put(ClickHouseSinkConnector.PORT, db.getFirstMappedPort().toString());

